text/0072-metric-observer.md

```diff
@@ -0,0 +1,202 @@
+`Monotonic` and `NonMonotonic` options, the same as Gauge instruments.
+
+Callbacks should avoid blocking.  The implementation may be required
+to cancel computation if the callback blocks for too long.
+
+Callbacks must not be called synchronously with application code via
+any OpenTelemetry API.  This prevents the application from potentially
+deadlocking itself by being called synchronously from its own thread.
+Implementations that cannot provide this guarantee should prefer not
+to implement Observer instrsuments.
+
+Callbacks may be called synchronously in the SDK on behalf of an
+exporter, provided it does not contradict the requirement above.
+
+Callbacks should avoid calling OpenTelemetry APIs other than the
+interface provided to `Observe()`.  This prevents the SDK from
+potentially deadlocking itself by being called synchronously from its
+own thread.  We recognize this may be impossible or expensive to
+enforce.  SDKs should document how they respond to such attempts at
+re-entry.
+
+### Observer calling conventions
+
+Observer callbacks accept a Result interface, which supports both
+bound and direct calls as follows.
+
+For a "direct" observation with a specific LabelSet, call the Result
+directly using `Result.Observe(value, LabelSet)` to report an
+observation.
+
+For a "bound" observation with using a bound observer instrument,
+first `Bind()` the instrument with a LabelSet, then call the bound
+instrument passing the result: `Result.ObserveBound(value,
+BoundInstrument)`.  Callbacks MUST use bound instruments corresponding the
+Observer instrument for which they are registered.  It is an error if
+`Result.ObserveBound(value, BoundInstrument)` is called for a Result
+corresponding to a different Observer instrument.
+
+If the language supports method overloading, it may use `Observe` for
+both calling conventions.
+
+Multiple observations are possible in a single callback invocation.
+Likewise, it is permissible to mix bound and direct observations in a
+single callback invocation.
+
+The Result passed to a callback should not be used outside the
+invocation to which it is passed.
+
+### Pseudocode
+
+An example of a bound call:
+
+```
+class YourClass {
+  private static final Meter meter = ...;
+  private static final ObserverDouble cpuTemp =
+      meter.observerDoubleBuilder("cpuTemp")
+           .withKeys("core")
+           .build();
+  private static final ObserverDouble[] cpuTempByCore = new ObserverDouble.Bound[NUM_CORES];
+
+  void init() {
+    for (int i = 0; i < NUM_CORES; i++) {
+        cpuTempByCore[i] = cpuTemp.Bind(meter.createLabelSet("core", i));
+    }
+
+    cpuTemp.setCallback(
+        new ObserverDouble.Callback<ObserverDouble.Result>() {
+          @Override
+          public void update(Result result) {
+            for (int i = 0; i < NUM_CORES; i++) {
+              result.ObserveBound(getCPUTemp(i), cpuTempByCore[i]);
+          }
+        });
+  }
+}
+```
```

Comment:
It's not clear from the text why the user shouldn't be able to set a callback on a bound instrument directly. This seems like a significant difference between `Set` and `Observe`.

Reply:
I tried to add a justification, but I feel about 50/50 on the pros/cons of the one-callback-per-instrument decision.

Reply:
Much clearer (even if it's a clear expression of ambiguity about the feature). Thanks!